### PR TITLE
added quarkeys z40 keyboard

### PR DESCRIPTION
**Pull request by trenthoh**
added the quarkeys z40

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/quarkeys/z40/config.h [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+#define MATRIX_ROW_PINS { E6, B3, C7, C6 }
+#define MATRIX_COL_PINS { F4, F5, F6, F7, B6, B5, B4, D7, D6, D0, B1, B0 }
+#define UNUSED_PINS
+
+/* COL2ROW, ROW2COL */
+#define DIODE_DIRECTION COL2ROW
+
+#define RGB_DI_PIN F1
+#ifdef RGB_DI_PIN
+#define RGBLIGHT_ANIMATIONS
```
Comment (drashna) [inline keyboards/quarkeys/z40/config.h:38]: Could you explicitly define which animations are enabled? 
```suggestion
#define RGBLIGHT_EFFECT_BREATHING
#define RGBLIGHT_EFFECT_RAINBOW_MOOD
#define RGBLIGHT_EFFECT_RAINBOW_SWIRL
#define RGBLIGHT_EFFECT_SNAKE
#define RGBLIGHT_EFFECT_KNIGHT
#define RGBLIGHT_EFFECT_CHRISTMAS
#define RGBLIGHT_EFFECT_STATIC_GRADIENT
#define RGBLIGHT_EFFECT_RGB_TEST
#define RGBLIGHT_EFFECT_ALTERNATING
#define RGBLIGHT_EFFECT_TWINKLE
```

File: keyboards/quarkeys/z40/readme.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
+# z40
+
+![z40](imgur.com image replace me!)
```
Comment (drashna) [inline keyboards/quarkeys/z40/readme.md:3]: Please fill out.

File: keyboards/quarkeys/z40/readme.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    make z40:default:flash
+
+See the [build environment setup](https://docs.qmk.fm/#/getting_started_build_tools) and the [make instructions](https://docs.qmk.fm/#/getting_started_make_guide) for more information. Brand new to QMK? Start with our [Complete Newbs Guide](https://docs.qmk.fm/#/newbs).
+
+## Bootloader
+
+Enter the bootloader in 3 ways:
+
+* **Bootmagic reset**: Hold down the key at (0,0) in the matrix (usually the top left key or Escape) and plug in the keyboard
+* **BootloadHID reset**: Hold down the key connected to the `A0` and `B0` pins on the MCU if it is known (often top left or bottom left) and plug in the keyboard
```
Comment (drashna) [inline keyboards/quarkeys/z40/readme.md:26]:
```suggestion
```

File: keyboards/quarkeys/z40/rules.mk [hunk 4]
```diff
@@ -0,0 +1,25 @@
+# MCU name
+MCU = atmega32u4
+
+# Bootloader selection
+#BOOTLOADER = atmel-dfu
+ARCH = AVR8
```
Comment (drashna) [inline keyboards/quarkeys/z40/rules.mk:6]:
```suggestion
BOOTLOADER = atmel-dfu
```

File: keyboards/quarkeys/z40/rules.mk [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+MOUSEKEY_ENABLE = yes       # Mouse keys
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = no         # Commands for debug and configuration
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = yes            # USB Nkey Rollover
+RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
+AUDIO_ENABLE = no           # Audio output
+
+KEY_LOCK_ENABLE = no        # Enable KC_LOCK support
```
Comment (drashna) [inline keyboards/quarkeys/z40/rules.mk:21]:
```suggestion
```

File: keyboards/quarkeys/z40/rules.mk [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+# Build Options
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = yes      # Enable Bootmagic Lite
+MOUSEKEY_ENABLE = yes       # Mouse keys
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = no         # Commands for debug and configuration
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
```
Comment (drashna) [inline keyboards/quarkeys/z40/rules.mk:16]:
```suggestion
```

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by trenthoh** [comment]
Keyboard creator asked for more changes involving lighting, default mapping and via.
Will re-do.

**Pull request closed without merging**